modules/core/src/smithy4s/schema/CollectionTag.scala

```diff
@@ -28,40 +30,101 @@ sealed trait CollectionTag[C[_]] {
       builder.result()
     }
   }
 
-  case object Vector extends CollectionTag[Vector] {
+  case object VectorTag extends CollectionTag[Vector] {
     override def name: String = "Vector"
     override def iterator[A](c: Vector[A]): Iterator[A] = c.iterator
 
     override def build[A](put: (A => Unit) => Unit): Vector[A] = {
-      val builder = cols.Vector.newBuilder[A]
+      val builder = Vector.newBuilder[A]
       put(builder.+=(_))
       builder.result()
     }
   }
+
+  case object IndexedSeqTag extends CollectionTag[IndexedSeq] {
+
+    override def name: String = "IndexedSeq"
+    override def iterator[A](c: IndexedSeq[A]): Iterator[A] = c.iterator
+
+    override def build[A](put: (A => Unit) => Unit): Vector[A] = {
+      val builder = Vector.newBuilder[A]
+      put(builder.+=(_))
+      builder.result()
+    }
+
+    /**
+      * Returns a builder that may be able to store the elements in an unboxed
+      * fashion
+      */
+    private[smithy4s] def compactBuilder[A](
+        schema: Schema[A]
+    ): ((A => Unit) => Unit) => IndexedSeq[A] =
+      schema.compile(CTSchemaVisitor) match {
+        case Some(ct) => { (put: (A => Unit) => Unit) =>
+          val builder = cols.ArraySeq.newBuilder(ct)
+          put(builder.+=(_))
+          builder.result()
+        }
+        case None => { (put: (A => Unit) => Unit) =>
+          val builder = IndexedSeq.newBuilder[A]
+          put(builder.+=(_))
+          builder.result()
+        }
+      }
+  }
+
+  private[this] type MaybeCT[A] = Option[ClassTag[A]]
+
+  /**
+   * Retrieves a ClassTag whenever possible.
+   */
+  // format: off
+  private[this] object CTSchemaVisitor extends SchemaVisitor[MaybeCT]{
+    val primitiveCT = Primitive.deriving[ClassTag]
+    def primitive[P](shapeId: ShapeId, hints: Hints, tag: Primitive[P]): MaybeCT[P] = Some(primitiveCT(tag))
+    def collection[C[_], A](shapeId: ShapeId, hints: Hints, tag: CollectionTag[C], member: Schema[A]): MaybeCT[C[A]] = tag match {
+      case ListTag => Some(implicitly[ClassTag[List[A]]])
+      case SetTag => Some(implicitly[ClassTag[Set[A]]])
+      case VectorTag => Some(implicitly[ClassTag[Vector[A]]])
+      case IndexedSeqTag => Some(implicitly[ClassTag[IndexedSeq[A]]])
+    }
+    def map[K, V](shapeId: ShapeId, hints: Hints, key: Schema[K], value: Schema[V]): MaybeCT[Map[K,V]] = Some(implicitly[ClassTag[Map[K, V]]])
+    def enumeration[E](shapeId: ShapeId, hints: Hints, values: List[EnumValue[E]], total: E => EnumValue[E]): MaybeCT[E] = None
+    def struct[S](shapeId: ShapeId, hints: Hints, fields: Vector[SchemaField[S, _]], make: IndexedSeq[Any] => S): MaybeCT[S] = None
+    def union[U](shapeId: ShapeId, hints: Hints, alternatives: Vector[SchemaAlt[U, _]], dispatch: U => Alt.SchemaAndValue[U, _]): MaybeCT[U] = None
+    def biject[A, B](schema: Schema[A], to: A => B, from: B => A): MaybeCT[B] = {
+      if (to.isInstanceOf[Newtype.Make[A, B]]) apply(schema).asInstanceOf[MaybeCT[B]]
```

Comment:
This is a hacky but when we have newtypes of primitives, we should attempt to retain compaction